### DISCUSSION TITLE
ReactDOM挂载的节点组件中使用useInjectable报错，应该如何处理

### DISCUSSION THREAD

**Discussion by lovelytan** [opensumi/core#1796] [Q&A] [answered]
需求：实现左下角登录功能。左下角自定义按钮，点击弹出全屏信息录入框，提交后更新左下角按钮状态。

1. 我编写了一个全屏信息录入表单组件，并使用二次封装的openSumi websocket通信服务
2. 在左下角tabbar注册了一个按钮及控制命令，利用when实现menuItem动态更新；
3. 点击按钮，控制命令中用ReactDOM渲染全屏组件
4. 页面报错，报错信息如下。封装的通信服务在openSumi实例中使用没有问题，将useInjectable那行注释可以正常渲染组件。
![image](https://user-images.githubusercontent.com/17985841/196175038-d25d78aa-6b11-4e78-9226-e3cb58fac8c3.png)

相关代码如下：

```typescript
// 组件代码
const Tpl: React.FC = () => {
    const {post} = useInjectable<IAPIService>(IAPIService);  // 将这行注释可以正常渲染组件
 
    async submit() {
        // await post()
        // todo
    }     

    return (
        <div>
        {/* 信息录入表单 */}
        </div>
    )
}

function show() {
    let container =  document.createElement('div');
    document.body.appendChild(container)
    // todo

    ReactDOM.render(<Tpl />, container)
}

export const myComponentFun = {
    show
}



//  控制代码
registerCommandsHandler(commands: CommandRegistry) {
    commands.registerCommand({id: 'xxx'}, {
        execute: () => {
            myComponentFun.show()
        }
    })
}
```

**Comment by erha19** [comment, marked as answer]
@naduohua 由于 `useInjectable` 是依赖挂载在例如 `#main` 节点上的 `ContextProvider` 去获取 DI 实例，你这里直接挂载到 `body->div` 上面是肯定不能用的。

你这里可以考虑直接将你的信息录入界面挂载到全局的某个常驻显示的视图模块下，例如 Toolbar 区域

或者考虑采用  `IDialogService` 里面的 Open 方法去挂载你的 React 组件
